docs/docs/spec.md

```diff
@@ -927,36 +927,28 @@ the output of the following operation:
+    > types stay non-null during substitution, even if they don't have an
+    > explicit `@NonNull` annotation on them.
     >
-    > Consider an unannotated user of `ImmutableList.Builder<Foo> builder`. Its
-    > type argument `Foo` will have a [nullness operator] of `UNSPECIFIED`.
-    > Without this special case, the parameter of `builder.add` would have a
-    > nullness operator of `UNSPECIFIED`, too. Then, when a lenient tool would
-    > check whether the [some-world] subtyping relation holds for
-    > `builder.add(null)`, it would find that it does.
+    > For an example of such a type, consider `Comparable`, a `@NullMarked`
+    > interface that declares a non-nullable type parameter `T` and a method
+    > `compare(T)`. By JSpecify rules, the method parameter has type `T
+    > NO_CHANGE`, and that type is null-exclusive under every parameterization
+    > in all worlds. Now consider a null-unmarked class that declares a method
+    > `Comparable<Foo> foo()`, which by JSpecify rules has a type argument `Foo
+    > UNSPECIFIED`. In this example, the question is what type
+    > `foo().compare(...)` accepts, and the answer changes as a result of this
+    > part of the subsitution rules:
     >
-    > To solve this, we need a special case for substitution for null-exclusive
-    > type parameters like the one on `ImmutableList.Builder`. That special case
-    > needs to produce a type with a nullness operator other than `UNSPECIFIED`.
-    > One valid option is to produce `NO_CHANGE`; we happened to choose
-    > `MINUS_NULL`.
-    >
-    > The choice between `NO_CHANGE` and `MINUS_NULL` makes little difference
-    > for the parameter types of `ImmutableList.Builder`, but it can matter more
-    > for other APIs' *return types*. For example, consider `@NullMarked class
-    > Foo<E extends @Nullable Object>`, which somewhere uses the type
-    > [`FluentIterable<E>`]. `FluentIterable` has a method `Optional<E>
-    > first()`. Even when `E` is a type like `String UNION_NULL` (or `String
-    > UNSPECIFIED`), we know that `first().get()` will never return `null`. To
-    > surface that information to tools, we need to define our substitution rule
-    > to return `E MINUS_NULL`: If we instead used `E NO_CHANGE`, then the
-    > return type would look like it might include `null`.
+    > -   If JSpecify were to directly subsitute `Foo UNSPECIFIED` in for `T`,
+    >     then the parameter type, which started out as non-null, would become
+    >     unspecified during subsitution. As a result, lenient checkers would
+    >     allow the call `foo().compare(null)`, since `Foo UNSPECIFIED` is
+    >     [null-inclusive under every parameterization] in [some world].
+    > -   To avoid that, JSpecify recognizes that the parameter is non-null, and
+    >     it performs substitution as if the parameter type were `T MINUS_NULL`
+    >     instead of `T NO_CHANGE`.
```

Comment:
I was tempted to add here: "JSpecify treats those types equivalently before substitution, so it makes sense for JSpecify to ensure that that are equivalent after substitution, too."

I couldn't decide whether that would make things better or worse.

Reply:
I had to think for a bit before I understood what you meant by "those types", so I don't think that sentence helps.

But maybe if you said more explicitly something like "JSpecify treats `T NO_CHANGE` just like `T MINUS_NULL` when `T` is non-nullable, and this rule preserves that equivalence through substitution, even when substituting `UNSPECIFIED` types."

If you do that, you should do it after these two bullets, not as part of the second bullet.